include/utilities/mdframe/dimension.hpp

```diff
@@ -0,0 +1,86 @@
+#ifndef NGEN_IO_MDFRAME_DIMENSION_HPP
+#define NGEN_IO_MDFRAME_DIMENSION_HPP
+
+#include <boost/optional.hpp>
+
+namespace ngen {
+
+class mdframe;
+
+namespace detail {
+
+/**
+ * Dimension Key
+ * 
+ * Provides a tagged dimension structure,
+ * with an optional size constraint.
+ */
+struct dimension {
+
+    struct hash
+    {
+        std::size_t operator()(const dimension& d) const noexcept
+        {
+            return std::hash<std::string>{}(d.m_name);
+        }
+
+        static std::size_t apply(const dimension& d) noexcept
+        {
+            return dimension::hash{}(d);
+        }
+    };
+
+    dimension()
+        : m_name()
+        , m_size() {};
+
+    dimension(const std::string& name)
+        : m_name(name)
+        , m_size(static_cast<std::size_t>(-1)) {};
+    
+    dimension(const std::string& name, std::size_t size)
+        : m_name(name)
+        , m_size(size) {};
+
+    dimension(const dimension& d)
+        : m_name(d.m_name)
+        , m_size(d.m_size) {};
+
+    dimension(dimension&& d)
+        : m_name(std::move(d.m_name))
+        , m_size(std::move(d.m_size)) {};
+
+    dimension& operator=(const dimension& d) {
+        this->m_name = d.m_name;
+        this->m_size = d.m_size;
+        return *this;
+    }
+
+    dimension& operator=(dimension&& d) {
+        this->m_name = std::move(d.m_name);
+        this->m_size = std::move(d.m_size);
+        return *this;
+    }
+    
+    bool operator==(const dimension& rhs) const
```

Comment:
Why compare hashes of `m_name` rather than `m_name` directly? The cost is actually higher, since the hash is reading the whole name and computing something from it.

Same applies to `variable` and anywhere else this pattern appears

Reply:
I think my initial thought was that the hash would be derived from not only `m_name`, but also the dimension size, but it probably makes more sense semantically to derive equality from only dimension names, since we don't want multiple dimensions with the same name in an `mdframe`.